### PR TITLE
Entry API, by-federates-with filter

### LINKED ISSUE #1894: The entry API should allow me to filter entries by federates with
As an operator, I would like to see all the entries that federate with a particular trust domain. This is particularly necessary when trying to delete or remove a trust bundle - are there any entries currently using it? The only way to do this today is to list _all_ entries, then iterate through them looking for the target trust domain name.

https://github.com/spiffe/spire/blob/bda8803756b29428506e66b20b12a65c58817ea0/proto/spire/api/server/entry/v1/entry.proto#L45-L49

### PR DESCRIPTION
**Pull request by lsantire**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**

- Entry API: ListEntries service
- Datastore: ListRegistrationEntries

**Description of change**

The change adds a new filter to the ListEntries service that allows callers to ask for all the entries that federate with a particular list of trust domains. Also, similarly to the BySelectors filter, the ByFederatesWith filter will try to match the given trust domains with the federated trust domains of each entry, following a "subset" or "exact" policy, according to the request.

**Which issue this PR fixes**

fixes #1894

### CODE REVIEW

**Comment by azdagron** [comment]
Thanks for opening this @Isantire! Many of us are heads down with Kubecon his week, so there might be some delay in taking a deep look at this PR.

**Review by azdagron — COMMENTED** [review]
Awesome, other than a few nits around naming and such I think this is good to go. Thanks, @lsantire!

File: proto/spire/types/federateswithmatch.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+option go_package = "github.com/spiffe/spire/proto/spire/types";
+
+message FederatesWithMatch {
+    enum MatchBehavior {
+        // Indicates that the federated trust domains in this match are 
+        // equal to the candidate trust domains, independent of ordering.
+        MATCH_EXACT = 0;
+
+        // Indicates that the federated trust domains in this match are 
+        // a subset of the candidate trust domains.
```
[the remark this answers is not in the document]
Comment (lsantire) [reply proto/spire/types/federateswithmatch.proto]: You are right, I just read it again and it was super confusing. I think your suggested comment is really clear.

**Comment by lsantire** [comment]
> Awesome, other than a few nits around naming and such I think this is good to go. Thanks, @lsantire!

Great @azdagron, thanks for reviewing this PR, I know it's not a short one.

After merging master into this branch and re-generating the .pb.go files, I noticed that many new changes appeared. I just used the _make generate_ command, then I checked that unit and integration tests were still passing and everything seems to work fine!

File: pkg/server/plugin/datastore/sql/sql_test.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/golang/protobuf/ptypes/wrappers"
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql_test.go:18]: The sql_test.go file (along with the rest of the codebase) has been ported to use the new protobuf library. The wrappers for those live in the "google.golang.org/protobuf/types/known/wrapperspb" package, which is imported below. We should use that package instead.

**Review by azdagron — COMMENTED** [review]
One small nit with an old protobuf library import being reintroduced into sql_test.go. After that I think we can get this in!

File: pkg/server/plugin/datastore/sql/sql_test.go (hunk 2 above)
Comment (lsantire) [reply pkg/server/plugin/datastore/sql/sql_test.go]: Good catch, thank you!

**Review by azdagron — APPROVED** [review]
Awesome, thanks @lsantire !

**Pull request merged**